core/src/main/java/org/elasticsearch/index/query/PercolateQueryBuilder.java

```diff
@@ -459,4 +472,83 @@ public Weight createNormalizedWeight(Query query, boolean needsScores) throws IO
         }
     }
 
+    private static PercolateQuery.QueryStore createStore(PercolatorFieldMapper.PercolatorFieldType fieldType,
+                                                         QueryShardContext context,
+                                                         boolean mapUnmappedFieldsAsString) {
+        return ctx -> {
+            LeafReader leafReader = ctx.reader();
+            BinaryDocValues binaryDocValues = leafReader.getBinaryDocValues(fieldType.getQueryBuilderFieldName());
+            Bits bits = leafReader.getDocsWithField(fieldType.getQueryBuilderFieldName());
+            return docId -> {
+                if (bits.get(docId)) {
+                    BytesRef qbSource = binaryDocValues.get(docId);
+                    if (qbSource.length > 0) {
+                        XContent xContent = PercolatorFieldMapper.QUERY_BUILDER_CONTENT_TYPE.xContent();
+                        try (XContentParser sourceParser = xContent.createParser(qbSource.bytes, qbSource.offset, qbSource.length)) {
+                            return parseQuery(context, mapUnmappedFieldsAsString, sourceParser);
+                        }
+                    } else {
+                        return null;
+                    }
+                } else {
+                    return null;
+                }
+            };
+        };
+    }
+
+    private static PercolateQuery.QueryStore createLegacyStore(QueryShardContext context, boolean mapUnmappedFieldsAsString) {
+        return ctx -> {
+            LeafReader leafReader = ctx.reader();
+            return docId -> {
+                LegacyQueryFieldVisitor visitor = new LegacyQueryFieldVisitor();
+                leafReader.document(docId, visitor);
+                try (XContentParser sourceParser = XContentHelper.createParser(visitor.source)) {
```

Comment:
can visitor.source be null? I guess this is not possible on a legacy index but maybe we should check and throw eg. an IllegalStateException with a helpful message if it is null?